sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java

```diff
@@ -0,0 +1,768 @@
+                          getTable(),
+                          getIntegrationName(),
+                          getStagingBucketName(),
+                          gcpTmpDirName,
+                          getSnowflakeService(),
+                          getSnowflakeCloudProvider())))
+              .apply(FileIO.matchAll())
+              .apply(FileIO.readMatches())
+              .apply(readFiles())
+              .apply(ParDo.of(new MapCsvToStringArrayFn()))
+              .apply(ParDo.of(new MapStringArrayToUserDataFn<>(getCsvMapper())));
+
+      output.setCoder(getCoder());
+
+      emptyCollection
+          .apply(Wait.on(output))
+          .apply(
+              ParDo.of(
+                  new CleanTmpFilesFromGcsFn(
+                      getStagingBucketName(), gcpTmpDirName, getSnowflakeCloudProvider())));
+
+      return output;
+    }
+
+    private String makeTmpDirName() {
+      return String.format(
+          "sf_copy_csv_%s_%s",
+          new SimpleDateFormat("yyyyMMdd_HHmmss").format(new Date()),
+          UUID.randomUUID().toString().subSequence(0, 8) // first 8 chars of UUID should be enough
+          );
+    }
+
+    private static class CopyIntoStageFn extends DoFn<Object, String> {
+      private final SerializableFunction<Void, DataSource> dataSourceProviderFn;
+      private final String query;
+      private final String table;
+      private final String integrationName;
+      private final String stagingBucketName;
+      private final String tmpDirName;
+      private final SnowflakeService snowflakeService;
+      private final SnowflakeCloudProvider cloudProvider;
+
+      private CopyIntoStageFn(
+          SerializableFunction<Void, DataSource> dataSourceProviderFn,
+          String query,
+          String table,
+          String integrationName,
+          String stagingBucketName,
+          String tmpDirName,
+          SnowflakeService snowflakeService,
+          SnowflakeCloudProvider cloudProvider) {
+        this.dataSourceProviderFn = dataSourceProviderFn;
+        this.query = query;
+        this.table = table;
+        this.integrationName = integrationName;
+        this.stagingBucketName = stagingBucketName;
+        this.tmpDirName = tmpDirName;
+        this.snowflakeService = snowflakeService;
+        this.cloudProvider = cloudProvider;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) throws Exception {
+        String stagingBucketDir = this.cloudProvider.formatCloudPath(stagingBucketName, tmpDirName);
+        String output =
+            snowflakeService.copyIntoStage(
+                dataSourceProviderFn,
+                query,
+                table,
+                integrationName,
+                stagingBucketDir,
+                tmpDirName,
+                this.cloudProvider);
+
+        context.output(output);
+      }
+    }
+
+    public static class MapCsvToStringArrayFn extends DoFn<String, String[]> {
```

Comment:
Is there a reason to introduce two transforms here instead of having a single transform that converts Strings to type T ?

Reply:
Yes, the story behind it is to make connector more user-friendly. 

We wanted to handle the CSV parsing for the user and left them only mapping from `Object[]` to `T`. 
Otherwise the user would have to do CSV line parsing on their own and face the same issues that we could fix for them (like having a comma in string) by setting proper delimiters and quoting when calling COPY statement. 

Do you think it's OK like this?

Reply:
Sounds good.